### PR TITLE
Feature gps snapshot

### PR DESCRIPTION
**Pull request by Bjerkk**
**Features added**
This pr adds functionality for the following functionality for the map:
- Initially using the network or GPS to place yourself on the map, posts it to db
- Every 5 seconds the map is updated with the position of all the users, including yourself
- The distance to the other users is calculated
- The other users' yaw is used to display their orientation, this does however look pretty confusing with the markers currently used for other users

**How to review**
- Open the map
- Locate the other users
- Wait 5 seconds
- Tap the other users to see the distance to them

**Most important files**
- MapsViewModel
- MapsActivity

**This solves the following trello tasks:**
[Trello task](https://trello.com/c/Ex4pR6Eb/218-update-map-with-db-locations)
[Trello task](https://trello.com/c/dG2wWCXu/216-gps-snapshotting-retrieve-gps-location-of-user)
[Trello task](https://trello.com/c/K7dW94Va/223-post-update-gps-snapshot-of-user-to-db)
[Trello task](https://trello.com/c/imdwHRcp/224-get-current-users-location-from-db-since-dr-is-calculated-server-side)
[Trello task](https://trello.com/c/SVqUXMMD/225-calculate-distance-to-other-markers-is-currently-hardcoded)

### CODE REVIEW

**Review by Sworgon — DISMISSED** [review]
I would like some more comments in the code not just before the methods. Otherwise, the implementation is very nice

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MainActivity.kt [hunk 1]
```diff
@@ -40,6 +55,15 @@ class MainActivity : AppCompatActivity() {
             val intent = Intent(this,MapsActivity::class.java)
             startActivity(intent)
         }
+        createLocationRequest()
+
+        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
+        /*
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MainActivity.kt]: Skal man ikke bare slette dette?

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MainActivity.kt [hunk 2]
```diff
@@ -1,32 +1,47 @@
...
+    lateinit var mFusedLocationClient: FusedLocationProviderClient
+    //private lateinit var gpsViewModel: GPSViewModel
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        //AndroidThreeTen.init(this)
         setContentView(R.layout.activity_main)
 
         //Checks locations permissions, which are necessary for
-        checkLocationPermissions()
+        //checkLocationPermissions()
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MainActivity.kt]: Skal den være kommenteret ud?

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MainActivity.kt [hunk 3]
```diff
@@ -157,6 +181,8 @@ class MainActivity : AppCompatActivity() {
         startActivity(discoverableIntent)
     }
 
+    //can remove this function
+    /*
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MainActivity.kt]: Skal denne så stadig være kommeret ud?

File: DeadCrumbs/app/src/main/java/dead/crumbs/data/MapsRepository.kt [hunk 4]
```diff
@@ -2,9 +2,15 @@ package dead.crumbs.data
 
 import dead.crumbs.data.DAO.Maps_DAO
 import dead.crumbs.data.DAO.RSSI_DAO
+import io.swagger.client.models.Location
 
 class MapsRepository private constructor(private val mapsDao: Maps_DAO) {
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/data/MapsRepository.kt]: kan vi have at DAO altid staves med stort?

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MainActivity.kt [hunk 5]
```diff
@@ -226,5 +252,32 @@ class MainActivity : AppCompatActivity() {
         Toast.makeText(this@MainActivity, "${rssi.target_mac_address}'s distance is\n $dist m", Toast.LENGTH_LONG).show()
     }
 
+    fun createLocationRequest() {
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MainActivity.kt:224]: Can we have 1 comment describing what this function does.

**Review by Sworgon — DISMISSED** [review]
Super cool, dudes

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsActivity.kt [hunk 6]
```diff
@@ -57,8 +63,25 @@ class MapsActivity : AppCompatActivity(), OnMapReadyCallback,
...
+                viewModel!!.updateMapPositions(context, this@MapsActivity)
+                handler.postDelayed(this, 5000)
+            }
+        }
+        handler.postDelayed(runnable, 5000)
+    }
+
+    override fun onPause() {
+        super.onPause()
+        //when leaving the map, the handlers callbacks (the runnable) is removed
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsActivity.kt:83]: Nice

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 7]
```diff
@@ -108,4 +145,159 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+        )
+    }
+
+    //check if gps or network is enabled
+    private fun isLocationEnabled(context: Context): Boolean {
+        val locationManager: LocationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
+        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
+    }
+
+    //check if gps or network is enabled
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt]: er det ikke bare network

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 8]
```diff
@@ -108,4 +145,159 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+    private fun requestPermissions(activity: Activity) {
+        ActivityCompat.requestPermissions(
+            activity,
+            arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION),
+            PERMISSION_ID
+        )
+    }
+
+    //check if gps or network is enabled
+    private fun isLocationEnabled(context: Context): Boolean {
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt:169]: er det ikke bare gps

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 9]
```diff
@@ -108,4 +145,159 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+    fun getGPSLocation(context: Context, activity: Activity): Location? {
+        val locMan : LocationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
+        var location: Location ?= null
+        if (checkPermissions(context)) {
+            if (isLocationEnabled(context)) {
+                location = locMan.getLastKnownLocation(LocationManager.GPS_PROVIDER)
+                if(location == null && isNetworkEnabled(context))
+                    location = locMan.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
+
+                val swaggerLocation : io.swagger.client.models.Location = io.swagger.client.models.
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt:195]: tror generelt ikke vi siger io.swagger osv, men blot includer location i toppen og så blot skriver Location, da vi ikke har andre Locations.

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 10]
```diff
@@ -108,4 +145,159 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+    @SuppressLint("MissingPermission")
+    fun getGPSLocation(context: Context, activity: Activity): Location? {
+        val locMan : LocationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
+        var location: Location ?= null
+        if (checkPermissions(context)) {
+            if (isLocationEnabled(context)) {
+                location = locMan.getLastKnownLocation(LocationManager.GPS_PROVIDER)
+                if(location == null && isNetworkEnabled(context))
+                    location = locMan.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
+
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt:189]: burde man ikke igen her tjekke om location er null? Ellers vil den da crashe, tror jeg nok.

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 11]
```diff
@@ -108,4 +145,159 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+                val swaggerLocation : io.swagger.client.models.Location = io.swagger.client.models.
+                    Location(loggedInUser, 0.0, Position("Point", arrayOf(location.latitude, location.longitude)), getDateTime())
+                postLocation(swaggerLocation)
+
+                return location
+            }
+        } else {
+            requestPermissions(activity)
+        }
+        throw Exception()
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt]: Kan vi have sådan en besked i denne exception.

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 12]
```diff
@@ -108,4 +145,159 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+        val locMan : LocationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
+        var location: Location ?= null
+        if (checkPermissions(context)) {
+            if (isLocationEnabled(context)) {
+                location = locMan.getLastKnownLocation(LocationManager.GPS_PROVIDER)
+                if(location == null && isNetworkEnabled(context))
+                    location = locMan.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
+
+                val swaggerLocation : io.swagger.client.models.Location = io.swagger.client.models.
+                    Location(loggedInUser, 0.0, Position("Point", arrayOf(location.latitude, location.longitude)), getDateTime())
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt:196]: Kan man adde en comment ifht at yaw til 0 ikke er et problem, da det bliver opdateret om 5 sekunder.

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 13]
```diff
@@ -27,43 +40,67 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+                for(user in users.value!!) {
+                    try {
+                        locList.add(getLocation(user.username))
+                    }
+                    catch (e: java.lang.Exception){
+                        Toast.makeText(context, e.message , Toast.LENGTH_LONG).show()
+                    }
+                }
+            //adding the meMarker to the map
+            val marker = map.addMarker(newMarker( loc = LatLng(loc!!.latitude, loc.longitude), name = "Me", icon = R.mipmap.my_picture))
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt]: en lang linje

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 14]
```diff
@@ -27,43 +40,67 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
             markerList.add(marker)
+            //assign meMarker for easier update of orientation
+            meMarker = marker
+
+            //Making all the other markers for the map
+            for (user in locList) {
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt]: Kan vi gøre sådan at den der iterations variable heddder location e.l.

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 15]
```diff
@@ -108,4 +145,159 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+                }
+            }
+
+            //assign meMarker for easier update of orientation
+            val newMarkerList = mutableListOf<Marker>()
+
+            //clear map to be able to update it with new values
+            map.clear()
+            //for loop used to find the current user
+            for (user in newLocations) {
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt:230]: Tror disse to for loops kan merges.

Comment (User12837985123) [reply DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt:230]: Alt dette og det på linje 235

File: DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt [hunk 16]
```diff
@@ -108,4 +145,159 @@ class MapsViewModel (private val mapsRepository: MapsRepository) : ViewModel(){
...
+                if(user.value!!.user_ref == loggedInUser){
+                   continue
+                }
+                else {
+                    val marker = map.addMarker(
+                        newMarker(
+                            LatLng(
+                                user.value!!.position.coordinates!![0],
+                                user.value!!.position.coordinates!![1]
+                            ), user.value!!.user_ref, distanceInM(ownLat, ownLong, user.value!!.position.coordinates!![0], user.value!!.position.coordinates!![1]),
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/ui/MapsViewModel.kt]: Man kan lige få IDE'en til at beutify dette, så det står bedre.

File: DeadCrumbs/app/src/main/java/dead/crumbs/utilities/InjectorUtils.kt [hunk 17]
```diff
@@ -24,9 +24,14 @@ object InjectorUtils {
         {
             // ViewModelFactory needs a repository, which in turn needs a DAO from a database
             // The whole dependency tree is constructed right here, in one place
-            val mapsRepository = MapsRepository.getInstance(FakeDatabase.getInstance().mapsDao)
+            val mapsRepository = MapsRepository.getInstance(Database.getInstance().mapsDao)
             singletonMapsViewModelFactory = SingletonMapsViewModelFactory(mapsRepository)
         }
         return singletonMapsViewModelFactory as SingletonMapsViewModelFactory
     }
+/*
```
Comment (User12837985123) [inline DeadCrumbs/app/src/main/java/dead/crumbs/utilities/InjectorUtils.kt]: Burde nok bare slette dette, da der ikke længere er noget som hedder LocationRepository

**Comment by User12837985123** [comment]
Jeg kom til at sige comment i stedet for request changes, så husk lige at folde alle kommentarerne ud :D

**Comment by User12837985123** [comment]
Ellers ser det nice ud 👯‍♂️

**Review by User12837985123 — APPROVED** [review]
🖕

**Review by Sworgon — APPROVED** [review]
Super cool, dudes!

**Pull request merged**